Guides visual redesign

jathayde opened the pull request:
<img width="957" alt="Screenshot 2024-03-18 at 11 12 26 AM" src="https://github.com/rails/rails/assets/2223/9da5e009-134e-4069-b79e-01582da998e2">


@AmandaPerino @rafaelfranca @carlosantoniodasilva @matthewd 

### Motivation / Background

This Pull Request has been created because of a project with the Rails Foundation to improve the visual presentation of the Rails Guides, bringing them more inline with the new marketing rupbyonrails.org site.

### Detail

The CSS has been recreated from SCSS files, generated into the final pieces. The stylesrc directory contains those commented source files. 

### Additional information

This currently uses CodeKit offline to process the SCSS. Ideally this could be part of the `rake guides:generate` task, or a new task. 

### Checklist

Before submitting the PR make sure the following are checked:

* [X] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [X] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [X] Tests are added or updated if you fix a bug or add a feature.
* [X] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

jathayde committed:
Touch icon/svg icon, html theme color

jathayde committed:
Icons and graphical elements

jathayde committed:
Initial typographic baseline and utility SCSS

jathayde committed:
Color variables

jathayde committed:
Initial refactor of interrupting blocks

jathayde committed:
Work in progress block

jathayde committed:
Mobile layout basics

jathayde committed:
Initial Guides readme

jathayde committed:
Adding .DS_Store to .gitignore

jathayde committed:
cleaning up def lists

jathayde committed:
New icon for chapters

jathayde committed:
COdeKit Config (Delete when going to prod)

jathayde committed:
Original guides output for reference (delete for prod)

jathayde committed:
Adding FAQs

jathayde committed:
Code container question placeholder

jathayde committed:
style build (temp)

jathayde committed:
Styling the numbers in Hn tags seperate from rest of inner_html

jathayde committed:
Question/FAQ block style basics

jathayde committed:
Code block and syntax highlighting

jathayde committed:
Header graphic elements

jathayde committed:
Desktop vs mobile two column setup

jathayde committed:
Tables basic

jathayde committed:
Footer

jathayde committed:
Fixing interstitial rendering not showing

jathayde committed:
Fixing some missing code highlight colors

jathayde committed:
Code in Hn’s - 3/4/5 fixes for same size

jathayde committed:
improving contrast in light mode (dark code)

jathayde committed:
Styling subcol to match mock (color/underline only on hover, correct padding on UL)

jathayde committed:
Shifting scroll to OL from entire subcol

jathayde committed:
Top nav and logo mark styled for mobile

jathayde committed:
back to top button

jathayde committed:
Mobile header styles

jathayde committed:
Fixing spacing on mobile above main

jathayde committed:
Desktop base styling

jathayde committed:
House icon in nav

jathayde committed:
dropdown guides

jathayde committed:
Updating styling on flyout (non bold)

jathayde committed:
Rails HR variation

jathayde committed:
Different style for HR

jathayde committed:
Visited color should stay red on H2

jathayde committed:
H2 sizing for Calibre x-height difference

jathayde committed:
Setting font-variant-numeric: lining-nums tabular-nums; on guides dropdown

jathayde committed:
Code block without the icon and padding reset

jathayde committed:
Chapter spacing fix on H3

jathayde committed:
Version switcher and JS to change on drop down onChange

jathayde committed:
Fixing regression on interstitial blocks padding

jathayde committed:
Changing function to Turbo.visit for version swap

jathayde committed:
Carlos’ refactor & improvements on version switcher Guides

jathayde committed:
Fixing guides kindle call out

jathayde committed:
Guides index link is now visible, allows click to close

jathayde committed:
Fixing some black on black text in highlights for HTML content

jathayde committed:
Misc cleanup, adding a rails HR above Feedback block

jathayde committed:
Changing nav per Basecamp discussion

jathayde committed:
Minor cleanup on includes, first pass on dark mode

jathayde committed:
More dark mode testing

jathayde committed:
White icons for dark mode

jathayde committed:
Fixing missing borders on interstitials

jathayde committed:
Setting max-width on the #guides fly out

jathayde committed:
Hiding kindle for edge

jathayde committed:
Fixing home icon to go to “rubyonrails.org” no www

jathayde committed:
Changing rails home to go to home of directory (version of guides) vs root of domain

jathayde committed:
columns in guides drop down all the same width

jathayde committed:
Making keywords higher contrast in dark mode

jathayde committed:
Updated README for Guides

jathayde committed:
Merge branch 'guides_rebuild'

jathayde committed:
Removing the old guides folder used for reference during dev

carlosantoniodasilva commented on guides/source/layout.html.erb:
```diff
@@ -39,59 +51,83 @@
+  </main>
 
-  <div id="container">
+  <section id="feedback">
```
I ran into a small issue with this new ID, and I'm not sure about calling this block "feedback". Yes, it holds that last "feedback" copy at the bottom, but it's where all the main content for the guide is (where we `yield`). It just so happens that every guide ends with the Feedback paragraphs.

I have a content blocker that removes comments and a bunch of other stuff from the page, and it turns out it removed the whole content because of its `#feedback` ID. Yes, it's kind of a problem with my content blocker in Safari, as it looks okay in Chrome, but there may be others out there doing the same. We'd be okay if just the actual "feedback" paragraphs were hidden by something like this (i.e. if they were wrapped in a `#feedback` tag), not the whole content.

<details><summary>This is what it looked like, for reference</summary>
<p>

![Screenshot 2024-03-18 at 11 02 14](https://github.com/rails/rails/assets/26328/4a300d42-2d41-4007-8a05-f47e1b167827)

</p>
</details> 

Do we need the ID for styling or anything else? Can we use another ID here?

jathayde replied:
I changed it back to `#container`

carlosantoniodasilva commented on guides/source/layout.html.erb:
```diff
@@ -1,31 +1,43 @@
-  <% if badge_version = @edge ? "edge" : @version %>
+  <%# Needs to be refactored into the dropdown concept - see mocks %>
+  <!--<% if badge_version = @edge ? "edge" : @version %>
   <div>
     <div id="version-badge"><%= badge_version %></div>
   </div>
-  <% end %>
-  <div id="topNav">
+  <% end %>-->
```
Let's remove these commented pieces as well.

carlosantoniodasilva commented on guides/source/layout.html.erb:
```diff
@@ -39,59 +51,83 @@
+        <ul class="nav">
+          <li><a class="nav-item" id="home_nav" href="https://rubyonrails.org/">Home</a></li>
+          <li class="guides-index guides-index-large">
+            <a href="index.html" id="guidesMenu" class="guides-index-item nav-item">Guides Index</a>
```
On larger screens this seems to be slightly off:

![Screenshot 2024-03-18 at 11 11 28](https://github.com/rails/rails/assets/26328/dcc81793-328c-48f5-b04b-4812ea79faf4)

This is on my external monitor, QHD resolution (2560 x 1440 pixels), both Safari and Chrome.

It also seems to swap to a 2-column list for the guides index, which makes the first column go all the way to the end and overflow as you reduce the height.

<details><summary>2-column overflow sample</summary>
<p>

![Screenshot 2024-03-18 at 11 14 26](https://github.com/rails/rails/assets/26328/3e31208b-7cc6-464d-854a-961d89894a80)

</p>
</details> 

Resizing the window decreasing height eventually swaps it to a 3-column layout, which makes all the content/links show properly. Can we keep the 3-column layout on larger versions?

jathayde replied:
<img width="2289" alt="Screenshot 2024-03-18 at 11 08 15 AM" src="https://github.com/rails/rails/assets/2223/fd035fff-91dd-4a49-b303-d0a8a669d854">

Forcing the height based on em - 3 column maintains on 2560x1440, even at full screen

carlosantoniodasilva replied:
Thanks, the 3-column layout seems to be working now.

I'm still seeing that bit of weirdness with the top of the guides index overlay though, on the larger screen. It's actually moving as you resize, causing it to move down enough that the menu elements under it show up:

https://github.com/rails/rails/assets/26328/fd74e232-7158-4f7c-8367-ee9740459187

jathayde replied:
Ah that's what `vw` will do for ya. Changed to explicit pixel values in c99d716b and it holds its place to the upper right.

carlosantoniodasilva reviewed (commented):
Hey @jathayde, really nice work on the new design!

Can you post some screenshot(s) and/or video(s) in the PR description? It helps for people to have a better idea of the changes.

I had some initial thoughts/suggestions below, I'll try to take another look at the generated guides today, but wanted to send what I had on a first quick pass. Let me know if you have any questions. Thanks!

jathayde committed:
Revert "Adding .DS_Store to .gitignore"

This reverts commit ed086291d879eb19a73624c454ec0b4cc681b164.

jathayde committed:
Fixing typo

jathayde committed:
Renaming GUIDES_README to README for auto-render in github

jathayde committed:
Deleting old CSS files

jathayde commented on guides/assets/images/hr_rails.svg:
I'll run image optim on the folder, but I believe this were hit on export with SVGO Compressor. I'll remove any unused (old) images here as well.

jathayde replied:
SVGs were not compressed but have now been. Removed anything that appeared old - did not edit folders as those seemed to be the inline images for content. 

<img width="761" alt="Screenshot 2024-03-18 at 10 45 14 AM" src="https://github.com/rails/rails/assets/2223/eea7215c-1c78-43db-9ef2-52b2cb7ec53e">

jathayde committed:
Removing old images, compressing new ones

jathayde committed:
Removing commented out code

jathayde committed:
Removing additional old CSS files

jathayde committed:
Merge branch 'main' of github.com:rails/rails

jathayde committed:
Removing old version banner

jathayde committed:
Forcing 3 column layout at large sizes

jathayde committed:
Changing “#feedback” to “#container” to prevent content blocking

jathayde commented:
@carlosantoniodasilva I believe I've addressed the initial items, save the built-in generator for SCSS => CSS, which I do agree should not rely on CodeKit, but use dart sass or something equivalent from the asset pipeline to process `style.scss` and `highlight.scss` across to their final forms.

jathayde committed:
SCSS processing as part of guides generation

jathayde committed:
Merge branch 'main' of github.com:rails/rails

jathayde committed:
Removing codekit config

jathayde committed:
Updating readme to remove codekit reference

carlosantoniodasilva commented on guides/assets/stylesrc/_main.scss:
```diff
@@ -0,0 +1,717 @@
+    }
+
+    &:visited {
+      color: $gray-600;
```
The current guides do not keep a different visited style as far as I can tell, I'm guessing this is intentional?

![Screenshot 2024-03-18 at 16 56 51](https://github.com/rails/rails/assets/26328/85a2a01d-c54d-4879-ba59-a2d2f39511ff)

(not suggesting I have a preference, just pointing out a difference I noticed)

jathayde replied:
Best practice would be to have a visited color so people know where they have been. Not a hill I'm planning to die on :) Would you all like to change it to be the same as `a:link`?

carlosantoniodasilva replied:
No strong opinions here, I'll let others comment as well. Maybe ask Amanda too.

AmandaPerino replied:
I've also gone back and forth on this a lot. After clicking around I noticed the gray and was mildly annoyed by it for aesthetic reasons, but of course I also understand the adage of 'this is how I know where I have been'. I'd be curious to ask a broader audience. Will tag you @jathayde.

matthewd replied:
Perhaps a much more subtle distinction would be a better fit?

As reference material that's expected to be repeatedly visited over the course of years, it seems unfortunate to over-emphasize visitedness, even if it is helpful for a theoretical new visitor who's trying to remember which single page they were reading the other day / is working their way through every guide in turn.

AmandaPerino replied:
Good point @matthewd. Maybe it's the jump from red to gray that sits strange with me. A subtle change would be great - those who need it can still see their path, but it won't stand out that much.

carlosantoniodasilva replied:
Yeah, a more subtle distinction could work well :+1:

jathayde replied:
Went with light for hover (5% lighter) and darker for visited (20% darker) in 3740edb9 and 3632e453

carlosantoniodasilva committed:
Use dartsass, generate files to the output dir

`dartsass-rails` is part of the Rails Gemfile, and thus should always
be available. Use `bundle exec` to make sure that's the version we're
using, always.

Generate the files directly to the output dir, so they don't go to the
source assets folder as intermediate files, otherwise we'd need to
explicitly tell git to ignore them. We don't need the intermediate files
stored there anyway, so let's just generate directly to the final output
folder, which is already git ignored.

Also remove `stylesrc` folder/files from the output folder, as it's not
necessary there.

Finally, remove the style and highlight css files, as they're going to
be generated from scss now.

carlosantoniodasilva commented on guides/rails_guides/generator.rb:
```diff
@@ -105,6 +106,10 @@ def select_only(guides)
 
+      def process_scss
+        system "sass ./assets/stylesrc/style.scss:./assets/stylesheets/style.css ./assets/stylesrc/highlight.scss:./assets/stylesheets/highlight.css"
+      end
```
I think this is simple enough and should work for now.

I had a few minor suggestions that I pushed as a PR to your branch: https://github.com/jathayde/rails/pull/1

carlosantoniodasilva commented on guides/assets/stylesrc/config.codekit3:
Do we still need this config file for codekit?

jathayde replied:
Thought I removed it and I added it to my global .gitignore. It should be gone per commit 595f4c3c

carlosantoniodasilva commented on guides/source/layout.html.erb:
```diff
@@ -39,59 +45,83 @@
+              <option value="https://guides.rubyonrails.org/v2.3">2.3</option>
+            </select>
+          </span>
+        </div>
```
I feel like the "Version: [ Edge ]" portion is a bit misaligned here, to a point that it bugs me just a little, but maybe it's just me... should these be aligned differently?

![Screenshot 2024-03-18 at 17 29 26](https://github.com/rails/rails/assets/26328/05832276-a23d-4f6d-a950-46ccdf551ebb)
![Screenshot 2024-03-18 at 17 30 09](https://github.com/rails/rails/assets/26328/262d3f3d-232c-4d20-a8b6-0feffd51770d)

jathayde replied:
baseline is south by ~2 pixels. I'll get it to lock. Might just be the difference in fonts, but I'll have it work visually.

jathayde replied:
<img width="337" alt="Screenshot 2024-03-18 at 9 13 32 PM" src="https://github.com/rails/rails/assets/2223/7136e7b6-b0b9-42bd-b977-076d57d0004a">

1px is all it took :)

jathayde replied:
Resolved in 3b329554

carlosantoniodasilva commented on guides/README.md:
```diff
@@ -0,0 +1,27 @@
+
+### Why do we include LTR and RTL?
+
+LTR/RTL (Left to right/right to left) is a layout change based on the nature of the language the site is being displayed in. Arabic and Farsi are two well known "RTL" languages. If the site is automatically translated, then the layout will adjust (mirror horizontally) to be more in line with the text.
```
Just tried RTL, and it causes the guide content to get under the nav on the right, and some of the boxes go over:

![Screenshot 2024-03-18 at 17 33 32](https://github.com/rails/rails/assets/26328/5699fd59-caf3-440a-9c69-14393266c5cc)

It also appears to create some horizontal scrolling, creating a lot of space between the content and the menu (home / guides index / contribute).

Are there plans to work on this support separately?

jathayde replied:
Yes. It's an item we deemed not critical for launch. It was intended for follow up later this week. Ideally with an arabic or farsi-speaking developer where we can check any issues that occur in a real RTL language.

carlosantoniodasilva commented on guides/source/layout.html.erb:
```diff
@@ -1,31 +1,37 @@
+  <nav id="topNav" aria-label="Secondary">
     <div class="wrapper">
       <strong class="more-info-label">More at <a href="https://rubyonrails.org/">rubyonrails.org:</a> </strong>
       <span class="red-button more-info-button">
```
Very minor: This is shown in smaller screens, but the cursor is text. It should instead be a cursor that indicates a  clickable element (even thought this is targeting more mobile screens). Hopefully this GIF explains it better:

![CleanShot 2024-03-18 at 17 39 34](https://github.com/rails/rails/assets/26328/b0fe614c-e4e1-4e39-ab49-88e750bae117)

jathayde replied:
Ah straightforward fix. Will switch cursor to `pointer`

jathayde replied:
Resolved in 3fb079a9

carlosantoniodasilva reviewed (commented):
@jathayde thanks for the updates! Here's a few more minor things I noticed while doing another pass / some more testing.

There's also a PR to your branch related to the sass generation, thanks for getting that working btw.

rafaelfranca commented on guides/rails_guides/generator.rb:
```diff
@@ -105,6 +106,10 @@ def select_only(guides)
       end
 
+      def process_scss
+        system "sass ./assets/stylesrc/style.scss:./assets/stylesheets/style.css ./assets/stylesrc/highlight.scss:./assets/stylesheets/highlight.css"
```
I don't think we have sass in the machine that generates the guides and I'd like to avoid installing another dependency. Do we need to run the sass given the css is also present in the repository?

rafaelfranca replied:
Ah, carlos's PR solve this problem.

rafaelfranca commented on guides/source/layout.html.erb:
```diff
@@ -39,59 +45,83 @@
+          <span id="version_switcher">
+            Version:
+            <select class="guides-version">
+              <option value="https://edgeguides.rubyonrails.org/" selected>Edge</option>
```
Why is Edge the selected? We don't want people to use the documentation of a version of Rails that wasn't released yet. Of couse it should be the selected based on which version we are on, so this should be dynamic somehow.

jathayde replied:
So on build, it builds edge. 7.1 doesn't exist on `rake guides:generate`. Is there a way to build 7.1 locally to test this?

@carlosantoniodasilva did your change to guides.js set this to update to whatever is active? I'm fine manually setting selected to 7.1 for now since loading the domain would load that, but ideally it works as @rafaelfranca describes and just selects which version it is on.

matthewd replied:
`@edge` / `@version` (from the currently removed version badge above) can probably help with this.

jathayde replied:
I'm sure there's a better way to do this, but my first "sledgehammer" way to get this in is a conditional on each `option`, e.g.:

```
<option value="https://edgeguides.rubyonrails.org/" <% if @edge == true %>selected<% end %> >Edge</option>
```

which seems dirty. Is there a better way? Maybe an interator over an array of versions that build each option?

matthewd replied:
Yeah, edge probably needs individual handling anyway, but for the rest, collapsing the full links into an array of version numbers seems both helpful and independently desirable :+1:

carlosantoniodasilva replied:
We don't need to refactor the list into Ruby generated list for now, as there's a similar list on the [intro for non-edge guides here](https://github.com/rails/rails/blob/94faed4dd4c915829afc2698e055a0265d9b5773/guides/source/_welcome.html.erb#L19-L32). We can circle back on refactoring those two lists later.

As for `selected`, we could do it in Ruby when the page is rendered via `@edge / @version`, but I thought we could do it in JS similar to how the current "guide" gets selected on the mobile version that shows a select:

https://github.com/rails/rails/blob/94faed4dd4c915829afc2698e055a0265d9b5773/guides/assets/javascripts/guides.js#L69

That's what I tried to accomplish in this bit of JS: https://github.com/rails/rails/pull/51341/files/bef423f59ec06f346ec64ac35a143f1dd7a9ecd9#diff-edcbfbf5f4b023773d5793ae6ac38eac5a5ec6213c25f0ca6b7d9a8e008621c7R67-R70

I think we could just remove the default `selected` from Edge here, if that JS works it should always show the current version based on the hostname/path.

jathayde replied:
So I think first is remove `selected` and let JS handle that, and second is add the version numbers to an array and build the option line programmatically (just static HTML, but easier editing in the future for adding/removing versions). Does that sound like a plan?

SleeplessByte replied:
https://github.com/rails/rails/pull/51341/files/bef423f59ec06f346ec64ac35a143f1dd7a9ecd9#diff-edcbfbf5f4b023773d5793ae6ac38eac5a5ec6213c25f0ca6b7d9a8e008621c7R72-R74 probably breaks arrow-key changing of the selector as it's unlikely the `visit` call will retain the focus. A list of links would probably be more appropriate.

That said, personally I don't like to defer to JavaScript to select the right option, if the right option is known when the HTML is generated.

carlosantoniodasilva replied:
@SleeplessByte it shouldn't break arrow navigation, since the "change" event only triggers when you actually select one (i.e. press enter on an option)

@rafaelfranca we do. I implemented that following the other select though, for the current guide, even thinking I could've reused the same "change" event... but I can change.

SleeplessByte replied:
@carlosantoniodasilva no that is a misconception. Tab to the `select` element and do not open it. Now press down arrow. This triggers a change event.

Reproducible for chromium based browsers and Firefox on Windows.

carlosantoniodasilva replied:
@SleeplessByte it might be and I just don't know about it 👍 . My quick test on Chrome / Safari on a Mac, even using tab + down, displays the select options:

<details><summary>Details</summary>
<p>

https://github.com/rails/rails/assets/26328/310acb38-a207-4c9f-8f97-d290d121e0da

</p>
</details>

carlosantoniodasilva replied:
Here's a suggestion on how to set the selected version upfront during guides generation, as well as using a list of versions to generate the options: https://github.com/jathayde/rails/pull/2

SleeplessByte replied:
@carlosantoniodasilva Ah, that explains where you are coming from. 
Here is a video to show the behaviour I described:

**Chrome**:

https://github.com/rails/rails/assets/1964376/383a348d-537d-4b5c-a092-bc2f8f2beafc

**Firefox**: 

https://github.com/rails/rails/assets/1964376/cd27bed9-49b6-4881-a54d-46e8ef388aaf



The keys I pressed are:
- Tab
- Tab
- Down arrow

carlosantoniodasilva replied:
@SleeplessByte thanks, yeah not sure why the behavior is so different. The same problem in theory exists on the guide selector on smaller screens, likely something we'd have to look into improving with other accessibility things.

SleeplessByte replied:
(either way, I think we oughtta ignore it for now and group with https://github.com/rails/rails/pull/51341#issuecomment-2007717282 like Amanda suggests!)

jathayde committed:
Merge pull request #1 from carlosantoniodasilva/jathayde-main

Use dartsass, generate files to the output dir

jathayde committed:
Cursor to pointer for mobile style more button

jathayde committed:
Fixing vertical alignment of baseline on version_switcher

jathayde committed:
Fixing guides position over slide

matthewd commented on guides/assets/stylesrc/style.scss:
```diff
@@ -0,0 +1,215 @@
+// February 2024
+//
+// John Athayde (https://www.johnathayde.com) for Meticulous
+// ---------------------------------------------------------------------------- */
```
This feels like a very point-in-time comment

jathayde replied:
Previous guides had date information. Do you want me to strike line 5?

jathayde replied:
Removed & switched back to created/modified info as previous master file had in a1831007

matthewd commented on guides/source/layout.html.erb:
```diff
@@ -1,31 +1,37 @@
   <meta name="viewport" content="width=device-width, initial-scale=1">
   <title><%= yield(:page_title) %></title>
+  <link rel="preconnect" href="https://rsms.me/">
+  <link rel="stylesheet" href="https://rsms.me/inter/inter.css">
```
Should we self-host this instead of making users ping a third party site?

jathayde replied:
We certainly can (https://rsms.me/inter/ is the source site where the link for download/install also lives).

jathayde replied:
Moved locally in 103f5724

matthewd commented on guides/source/layout.html.erb:
```diff
@@ -39,59 +45,83 @@
-              <% documents_by_section.each do |section| %>
-                <div class="guides-section">
-                  <dt><%= section['name'] %></dt>
+      <nav id="feature_nav"> <%# we use nav as an element to also provide the aria role and improve semantic meaning. %>
```
This also feels more like a comment on the change than on the resulting code state

jathayde replied:
Combination intent:
1. Education: The guides were all div-itis, so encouraging semantic markup wherever possible is an educational opportunity to developers who have minimal exposure to HTML & CSS.
2. ARIA note so we don't end up with redundant `role="navigation"` call here.

jathayde replied:
Removed for now in 3632e453

matthewd commented on guides/source/layout.html.erb:
```diff
@@ -39,59 +45,83 @@
+      <nav id="feature_nav"> <%# we use nav as an element to also provide the aria role and improve semantic meaning. %>
+        <div class="header-logo">
+          <a href="index.html" title="Return to home page">Guides</a>
+          <span id="version_switcher">
```
We probably need to provide a way to detect that we're doing a public-site build, and only include the version switcher in that case. If someone is looking at it locally, I think it's clearer to only show their current version?

jathayde replied:
Since we've removed the upper right corner banner, would we want to show that version number here in lieu of the switcher? How would we detect between a development build (where we want to see it), vs a general local build, vs production build?

jathayde replied:
Improved to be explicit in 7e40297a

dhh replied:
Don't think we need to show anything there if there are no versions to select between. If you're looking at it locally, you know what you built from.

jathayde replied:
I took a stab at this and was not able to sort it out (passing an ENV variable I could get to generator, but not then check for that in the layout.html.erb. Just a bit beyond my ability to heal.)

AmandaPerino replied:
Hi all, this is a blocker for us until we get some advice on how to implement here. Is this a big enough issue to stop this going live, or can we tackle this after? We have a small list of silo'd issues we will work on once the bulk of the work is out the door, and can add this to the list. cc: @matthewd @carlosantoniodasilva

carlosantoniodasilva replied:
I am not sure there's much of a reason to _hide_ it locally and only show in a "production" build (edgeguides/guides), that would only make it harder to work on it locally (if you're working on the design for instance, you need a way to _see_ it so you can style it properly, etc.), since you'd now need to know that there's this element that only shows up in production builds, and how to build production vs non-production output for instance.

If you're building with a version different than Edge, it'd show that version as the currently one selected (with [my latest change](https://github.com/jathayde/rails/pull/2)), so it'd be similar to the previous "banner". Now you just get the other versions listed there, something we previously didn't have, and when selected they will just redirect you to the respective guides/edgeguides, even locally. That seems fine.

Right now I don't think we have any differentiation of building local vs building production, we just generate a bunch of static HTML to an output folder. Adding an extra "production vs local" variable just for hiding the version switcher (or hiding the other versions in there) seems to complicate it more.

matthewd commented on guides/source/layout.html.erb:
```diff
@@ -39,59 +45,83 @@
-                  <dt><%= section['name'] %></dt>
+      <nav id="feature_nav"> <%# we use nav as an element to also provide the aria role and improve semantic meaning. %>
+        <div class="header-logo">
+          <a href="index.html" title="Return to home page">Guides</a>
```
This is [already] a bad title: "home page" sounds more like rubyonrails.org than the guides index/list/catalog/home. Perhaps we should fix it while we're here?

jathayde replied:
Improved to be explicit in 7e40297a

matthewd commented on guides/README.md:
```diff
@@ -0,0 +1,27 @@
+
+To generate new guides into static files, type `rake guides:generate` from inside the `guides` folder. If you make changes to the HTML or ERB, you'll need to remove the "output" directory before running this command. The master SCSS files (style.scss, highlight.scss) will compile as part of this process. 
+
+## FAQ
```
I think there's some useful context in here, but this isn't the right place for it to live: future visitors would expect `guides/README` to be a readme about the content of the guides, not the more rarely edited styles.

jathayde replied:
Okay. There was no guides README before (it had been left behind with a note that it had been removed). I put this in as it took me a bit of digging to figure out how to build the guides the first time. I also wanted to explain some things to any who might want to contribute to the guides in the future. I expect that the content team would add additional detail here (w/r/t content of the guides) in their PR, but open to changes here.

carlosantoniodasilva replied:
I figured we'd add more to the readme about the guides themselves later on as well.

SleeplessByte commented:
I found some preliminary issues that should be relatively fast to fix -- but these can be fixed post-deploy as well. I would personally love to see these implemented, and we'll do a run with assistive technology later to uncover further issues.

## Horizontal scrollbar on Windows

This may be an `sdoc` upstream issue.

On windows you'll get a horizontal scrollbar because scrollbars on windows are by default not `overlay`. This is caused by `100vw` on the page footer, which seems to be an inherited style. Switching to `100%` fixes the issue. 

I have captured both the visual issue and the style applied that causes this issue.

| Visual issue | Cause |
| ------------ | ----- |
| ![Horizontal scrollbar on all pages](https://github.com/rails/rails/assets/1964376/261617ee-00e0-4445-b6a3-b5e1b3e86824) | ![Style declaration for the page footer](https://github.com/rails/rails/assets/1964376/2dda4288-32d1-4dda-b4bd-f635bd99b686) |

## Dark mode heading and links are low contrast

The headings and links are difficult to read for me personally, and upon checking that is because the contrast actually is very low.

| Visual issue | Cause |
| ------------ | ----- |
| ![Heading and link in dark mode](https://github.com/rails/rails/assets/1964376/5d812ec9-c1fd-45de-a8f3-15cafda1551a) | ![Contrast on header of 2.1](https://github.com/rails/rails/assets/1964376/6e162f90-a0c0-47d0-8837-f19e371c849f) |

Should aim for at least an AA rating which is 4.5:1. One way to accomplish this would be to [make both the background darker and the heading and links lighter](https://webaim.org/resources/contrastchecker/?fcolor=EA2E31&bcolor=0F0F0F) but it's very hard to keep the "Rails red" brand color.

The same is true for dark mode links inside information boxes [are low contrast](https://webaim.org/resources/contrastchecker/?fcolor=DF161B&bcolor=00F0FF).


## The `main` content isn't inside the `main` landmark

| Indication | Cause |
| ------------ | ----- |
| ![Visually indicating where the main content ends](https://github.com/rails/rails/assets/1964376/2397f153-4896-40a9-9801-016d8a221f31) | ![Landmark main ends before the actual guide content](https://github.com/rails/rails/assets/1964376/09be4bcc-cdd2-4aee-b47d-b1b860eb60e1) |

When I am using a screenreader, I often browse by landmark and headings. The main content of the guide isn't instead the main landmark. This isn't completely breaking my browsing experience, but it's probably not what was intended. The main content for the guide should be inside the `main` landmark.

## The table of contents column sidebar should probably be a `nav`. 

It is meant to navigate on the same page. In order to make that work, it should have an _accessible label_. There is already a descriptive heading inside of it (faulty `h3`, should be `h2` to ensure the heading markup is consecutive), so adding `aria-labelledby` and a unique `id` to the sidebar should suffice.

![image](https://github.com/rails/rails/assets/1964376/12488799-9ba3-47f2-bd74-589014a936ed)
 
## The _back to top_ button is not accessible

![The back to top button](https://github.com/rails/rails/assets/1964376/3172a1ee-8b2e-4164-876a-16aa6dd0e784)

- It is missing an accessible label. One could be added by adding `sr-only` / `visually-hidden` content.
- It links to `#`, but personally I would rather see it go to `main` which has the id `feature`.
- It always scrolls smoothly, which shouldn't happen if there is a preference for reduced motion. The fix for that is as follows
  ```javascript
  // https://caniuse.com/?search=matchMedia IE 10+; may want to test for `matchMedia` existing on `window` if IE 9 should be supported
  var mediaQuery = window.matchMedia('(prefers-reduced-motion: reduce)')
  var behaviour = mediaQuery.matches ? 'auto' : 'smooth'
  // now scroll
  ```

## Printing removes the code highlighting

Because of an aggressive rule in `boilerplate.scss`, when printing, all highlighting is lost. This is not great if we want to do some paper annotations. 

_Aside_: hide the `copy` buttons when printing.

| Visual issue | Cause |
| ------------ | ----- |
| ![Print stylesheet removes highlight](https://github.com/rails/rails/assets/1964376/39e06f5e-752f-4636-8ffe-632b296368d4) | | ![CSS removing all color when printing](https://github.com/rails/rails/assets/1964376/94c2faa9-0508-48dc-aa71-3caffa75ad9e) |
  
----------------------------------

A personal note: the light code blocks in dark mode are jarring, as are the _warning (yellow) boxes. I'd probably keep the dark code blocks for dark mode. This is a personal preference, but I found it hard to browse the code examples when tired.

AmandaPerino commented:
> I found some preliminary issues that should be relatively fast to fix -- but these can be fixed post-deploy as well. I would personally love to see these implemented, and we'll do a run with assistive technology later to uncover further issues.
> 

Thanks @SleeplessByte! Let's get this into another PR that focuses on accessibility once all of these current changes are pushed live.

carlosantoniodasilva committed:
Set the selected Rails version based on the generated guides version

Instead of setting it dynamically via JS, use the known version that the
guide is being generated with, or edge, to determine the selected Rails
version in the version dropdown.

Since each version is set when generating the guides, we can "hardcode"
the selected version upon guides generation, instead of using JS to set
it dynamically.

jathayde committed:
Merge pull request #2 from carlosantoniodasilva/jathayde-main

Set the selected Rails version based on the generated guides version

jathayde committed:
Moving Inter font from CDN to local

jathayde committed:
Merge branch 'main' of github.com:jathayde/rails

jathayde committed:
Improving the title attribute for home link/icon

jathayde committed:
Removing “moment in time” and mimicing old datestamps

jathayde committed:
Fixing horizontal scroll error

jathayde committed:
Minor changes to HTML entites as suggested

jathayde commented:
@SleeplessByte I fixed the footer 100vw => 100%, thanks for that catch. I switched main to the main yield area and changed subCol to nav.

We do have a follow on task to revisit print CSS entirely.

jathayde committed:
Changing visited link to be close to red.

jathayde committed:
more contrast between regular and dark(er) for links

jathayde committed:
Striking the ERB comment in layout about nav and ARIA

carlosantoniodasilva commented on guides/assets/fonts/inter.css:
```diff
@@ -0,0 +1,57 @@
+@font-face { font-family: "InterDisplay"; font-style: normal; font-weight: 800; font-display: swap; src: url("InterDisplay-ExtraBold.woff2") format("woff2"); }
+@font-face { font-family: "InterDisplay"; font-style: italic; font-weight: 800; font-display: swap; src: url("InterDisplay-ExtraBoldItalic.woff2") format("woff2"); }
+@font-face { font-family: "InterDisplay"; font-style: normal; font-weight: 900; font-display: swap; src: url("InterDisplay-Black.woff2") format("woff2"); }
+@font-face { font-family: "InterDisplay"; font-style: italic; font-weight: 900; font-display: swap; src: url("InterDisplay-BlackItalic.woff2") format("woff2"); }
```
That's a lot of fonts :), but it looks like we aren't using all of them... would it make sense to "simplify" this a bit, and only keep the variations we're using?

For example, I don't see "InterDisplay" used, but I see "Inter" and "InterVariable" (just don't know which variations of those we're using specifically)...

It also seems we're not referencing this inter.css file directly, but have these font-face declarations also on our style, can we keep just one or the other?

jathayde replied:
`InterVariable` is preferred. `Inter` and the static files are for older browser support. Will remove display weights.

jathayde replied:
```
  :root {
    font-family: Inter, sans-serif;
    font-feature-settings: 'liga' 1, 'calt' 1; /* fix for Chrome */
  }
  @supports (font-variation-settings: normal) {
    :root { font-family: InterVariable, sans-serif; }
  }
```
from `_main.scss` calls both. Have removed the rest in 19877ca7

carlosantoniodasilva commented on guides/assets/stylesrc/config.codekit3:
We may have forgotten to remove other codekit3 config file

carlosantoniodasilva commented on guides/assets/stylesrc/components/_code-container.scss:
```diff
@@ -0,0 +1,97 @@
+    padding-top: 1em !important;
+  }
+
+  &.important {} // refrenced in old code - actually used?
+  &.warning {} // referenced in old code - actually used?
```
Do we need to keep these?

jathayde replied:
I don't believe so. Removed in 9d4ea2bb

carlosantoniodasilva commented on guides/source/layout.html.erb:
```diff
@@ -39,59 +43,72 @@
+        </div>
+        <ul class="nav">
+          <li><a class="nav-item" id="home_nav" href="https://rubyonrails.org/">Home</a></li>
+          <li class="guides-index guides-index-large">
```
Not sure how much control we have over the columns here, I noticed on Safari the "Policies" group/guides are kept on the middle column, which causes it to go all the way to the bottom of the container, and at a certain size one guide breaks the line and makes the column overflow a bit. It doesn't seem to be a problem on Chrome because that same group is moved to the 3rd column, so while there's a bit of line breaking and/or height adjusting as you resize, it doesn't make it overflow.

<details><summary>Safari</summary>
<p>

https://github.com/rails/rails/assets/26328/9e5f85f5-1b59-4fc6-a91f-7f6ced67c048

</p>
</details> 


<details><summary>Chrome</summary>
<p>

https://github.com/rails/rails/assets/26328/57e5eb2e-89dc-47f6-a50a-db88c088201e

</p>
</details> 

Just noting it here, in case there's something you think we can do to prevent it. Not a blocker.

jathayde replied:
Reducing the guides-section-container height to 53em from 55em kicks policies to the next column for safari. 81277220

carlosantoniodasilva commented on guides/rails_guides/markdown.rb:
```diff
@@ -148,10 +148,20 @@ def generate_index
-            <h3 class="chapter"><img src="images/chapters_icon.gif" alt="" />Chapters</h3>
+          <nav id="subCol">
+            <h3 class="chapter">
+            <picture>
+            <!-- Using the `source`  HTML tag to set the dark theme image -->
+            <source
+              srcset="images/icon_book-close-bookmark-1-wht.svg"
+              media="(prefers-color-scheme: dark)"
+            />
+            <img src="images/icon_book-close-bookmark-1.svg" alt="Chapter Icon" />
+          </picture>
```
Nitpicking: indent here:

```suggestion
              <picture>
                <!-- Using the `source`  HTML tag to set the dark theme image -->
                <source
                  srcset="images/icon_book-close-bookmark-1-wht.svg"
                  media="(prefers-color-scheme: dark)"
                />
                <img src="images/icon_book-close-bookmark-1.svg" alt="Chapter Icon" />
              </picture>
```

jathayde replied:
Fixed in 29466016

carlosantoniodasilva commented on guides/assets/images/icon_file-code.svg:
Looks like we decided not to use this one, can be removed?

jathayde replied:
Removed in 7fe9952d

carlosantoniodasilva commented on guides/assets/images/rubyline.svg:
I couldn't find this one referenced anywhere, can be removed?

jathayde replied:
Yes, this was replaced by the rails HR line SVG. Removed in 7fe9952d

carlosantoniodasilva reviewed (commented):
I noticed a couple other minor things doing one last pass, also had a comment about potentially cleaning up some of the fonts since we don't seem to be using them all... other than that, looks good to me.

(Note: I am aware we'll be doing some follow-up / a second pass on other style things too)

jathayde committed:
Removing codekit config again

jathayde committed:
Removing Inter Dislpay and extra css lines

jathayde committed:
Removing unused selectors in interstitial container

jathayde committed:
Moving policies to the next line, removing unused code in #guides

jathayde committed:
Cleaning up indent on picture declaration

jathayde committed:
Removing additional unused SVGs

jathayde commented:
@carlosantoniodasilva Thanks! I've addressed those items. Please let me know if there are any others.

jathayde committed:
Moving fonts to marcom site
